sklearn/cluster/tests/test_optics.py

```diff
@@ -0,0 +1,386 @@
+    n_points_per_cluster = 250
+
+    C1 = [-5, -2] + .8 * rng.randn(n_points_per_cluster, 2)
+    C2 = [4, -1] + .1 * rng.randn(n_points_per_cluster, 2)
+    C3 = [1, -2] + .2 * rng.randn(n_points_per_cluster, 2)
+    C4 = [-2, 3] + .3 * rng.randn(n_points_per_cluster, 2)
+    C5 = [3, -2] + 1.6 * rng.randn(n_points_per_cluster, 2)
+    C6 = [5, 6] + 2 * rng.randn(n_points_per_cluster, 2)
+    X = np.vstack((C1, C2, C3, C4, C5, C6))
+
+    # Compute OPTICS
+
+    clust = OPTICS(min_samples=9)
+
+    # Run the fit
+    clust.fit(X)
+
+    assert_equal(len(set(clust.labels_)), 6)
+
+
+def test_cluster_pruning():
+    # Tests pruning left and right, deletion of empty node_lists
+    # Parameters chosen specifically for this task
+
+    # Three pseudo clusters, 2 of which are too small
+    reach = np.array([np.inf, 0.9, 0.9, 1.0, 0.89, 0.88, 10, .9, .9, .9,
+                      10, 0.9, 0.9, 0.89, 0.88, 10, .9, .9, .9, .9])
+    # Normalize
+    reach = reach / np.max(reach[1:])
+
+    ordering = np.r_[0:20]
+    cluster_boundaries = _find_local_maxima(reach, 5)
+    root = _TreeNode(ordering, 0, 20, None)
+
+    # Build cluster tree inplace on root node
+    _cluster_tree(root, None, _find_local_maxima(reach, 5),
+                  reach, ordering, 5, .75, .7, .4, .3)
+    assert_equal(root.split_point, cluster_boundaries[0])
+    assert_equal(2, len(root.children))
+
+
+def test_sigmin_pruning():
+    # Tests pruning of insignificant splitpoints
+    # Parameters chosen specifically for this task
+
+    # Two pseudo clusters, 1 of which are too small
+    reach = np.array([np.inf, 0.9, 0.9, 0.9, 0.89, 0.88, 10, .9, .9, .9,
+                      10, 0.9, 0.9, 0.89, 0.88, 100, .9, .9, .9, .9])
+    # Normalize
+    reach = reach / np.max(reach[1:])
+
+    ordering = np.r_[0:20]
+    cluster_boundaries = _find_local_maxima(reach, 5)
+    root = _TreeNode(ordering, 0, 20, None)
+
+    # Build cluster tree inplace on root node
+    _cluster_tree(root, None, _find_local_maxima(reach, 5),
+                  reach, ordering, 5, .75, .7, .4, .3)
+    assert_equal(root.split_point, cluster_boundaries[0])
+    assert_array_equal(np.r_[0:15], root.children[0].points)
+
+
+def test_reach_dists():
+    # Tests against known extraction array
+
+    rng = np.random.RandomState(0)
+    n_points_per_cluster = 250
+
+    C1 = [-5, -2] + .8 * rng.randn(n_points_per_cluster, 2)
+    C2 = [4, -1] + .1 * rng.randn(n_points_per_cluster, 2)
+    C3 = [1, -2] + .2 * rng.randn(n_points_per_cluster, 2)
+    C4 = [-2, 3] + .3 * rng.randn(n_points_per_cluster, 2)
+    C5 = [3, -2] + 1.6 * rng.randn(n_points_per_cluster, 2)
+    C6 = [5, 6] + 2 * rng.randn(n_points_per_cluster, 2)
+    X = np.vstack((C1, C2, C3, C4, C5, C6))
+
+    # Compute OPTICS
+
+    clust = OPTICS(min_samples=10, metric='minkowski')
+
+    # Run the fit
+    clust.fit(X)
+
+    # Expected values, matches 'RD' results from:
+    # http://chemometria.us.edu.pl/download/optics.py
+    # Skip to line 385
```

Comment:
I don't think a line reference is a great idea!

Reply:
removed.